File: Development/getStructure.js
Prompt for AI:
Task: Create a Python script that monitors the current file structure of a project directory and logs changes such as new files, moved files, and deleted files in a user-friendly way. The script should update a currentFileStructure.txt file with the latest file structure tree but should not print the entire structure in the terminal. Instead, the terminal should only display concise updates, such as:

NewFile: "filepath"
MovedFile: "filepath"
DeletedFile: "filepath"
Requirements:

File Structure Monitoring:

The script should continuously monitor the specified project directory for changes in the file structure.
It should detect new, moved, or deleted files and directories.
Logging Updates in the Terminal:

The script should log only updates to the terminal, not the full file structure tree.
Logs should clearly state the type of change (NewFile, MovedFile, or DeletedFile) and the associated filepath.
Updating the File Structure File:

The script should maintain and update a currentFileStructure.txt file that contains the full file structure tree of the monitored directory.
The file structure should be updated each time a change is detected.
Implementation Language:

Use Python with libraries that suit React and Django project environments, such as os, pathlib, and watchdog for monitoring file changes.
Performance Considerations:

Ensure the script is efficient, with minimal resource usage when monitoring large projects.
Use optimized methods to check and update the file structure without excessive re-processing.
Sample Script:

Below is a sample script structure to help the AI generate the Python code based on the new requirements.
Sample Python Script:
python
Copy code
import os
import time
from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler
from pathlib import Path

# Define the path to the file structure file
FILE_STRUCTURE_PATH = 'currentFileStructure.txt'

class FileStructureHandler(FileSystemEventHandler):
    def __init__(self, root_dir):
        self.root_dir = root_dir
        self.previous_structure = self.read_file_structure()

    def on_any_event(self, event):
        # Check the event type and log the appropriate message
        if event.is_directory:
            return  # Skip directory events

        if event.event_type == 'created':
            print(f'NewFile: "{event.src_path}"')
        elif event.event_type == 'moved':
            print(f'MovedFile: "{event.dest_path}"')
        elif event.event_type == 'deleted':
            print(f'DeletedFile: "{event.src_path}"')

        # Update the file structure after handling the event
        self.update_file_structure()

    def read_file_structure(self):
        # Read the existing file structure from the file if it exists
        if not os.path.exists(FILE_STRUCTURE_PATH):
            return {}
        with open(FILE_STRUCTURE_PATH, 'r') as file:
            return {line.strip() for line in file.readlines()}

    def update_file_structure(self):
        # Generate the current file structure as a tree
        new_structure = self.get_directory_tree(self.root_dir)
        # Write the updated file structure to the file
        with open(FILE_STRUCTURE_PATH, 'w') as file:
            file.write(new_structure)

    def get_directory_tree(self, root_dir):
        # Recursively generate the directory tree as a string
        structure = ""
        for root, dirs, files in os.walk(root_dir):
            level = root.replace(root_dir, "").count(os.sep)
            indent = " " * 4 * (level)
            structure += f"{indent}{os.path.basename(root)}/\n"
            subindent = " " * 4 * (level + 1)
            for f in files:
                structure += f"{subindent}{f}\n"
        return structure


if __name__ == "__main__":
    # Define the root directory to monitor
    ROOT_DIR = Path('.').resolve()

    # Initialize the event handler and observer
    event_handler = FileStructureHandler(ROOT_DIR)
    observer = Observer()
    observer.schedule(event_handler, str(ROOT_DIR), recursive=True)

    # Start observing the file structure
    observer.start()
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        observer.stop()
    observer.join()
Example Execution:

Place the script in your project directory.
Run the script using Python: python monitor_structure.py.
The script will update the currentFileStructure.txt with the latest file structure tree and print concise updates to the terminal for any changes.
Suggestions for the AI:

The AI should refine the script to ensure it works seamlessly with React and Django environments, focusing on efficiency and the ability to handle large file structures.
Ensure that the logging and file updating processes are optimized and that the terminal output is clear and minimal as requested.
Let me know if you need any further adjustments or specific enhancements to this approach!

File: Development/getStructure.py
import os
import filecmp
import json
from datetime import datetime

# Define the root directory and the output file
ROOT_DIR = 'C:/Users/Thomas Ferrier/flyright-ai/Development'
OUTPUT_FILE = os.path.join(ROOT_DIR, 'currentFileStructure.md')
LOG_FILE = os.path.join(ROOT_DIR, 'file_changes.log')
PREVIOUS_STRUCTURE_FILE = os.path.join(ROOT_DIR, 'previousFileStructure.json')

def get_file_structure(root_dir):
    file_structure = {}
    for root, dirs, files in os.walk(root_dir):
        relative_root = os.path.relpath(root, root_dir)
        file_structure[relative_root] = files
    return file_structure

def save_file_structure(file_structure, output_file):
    with open(output_file, 'w') as f:
        for dir_path, files in file_structure.items():
            f.write(f"{dir_path}/\n")
            for file in files:
                f.write(f"  {file}\n")

def log_changes(new_structure, old_structure, log_file):
    new_files = []
    moved_files = []

    for dir_path, files in new_structure.items():
        if dir_path not in old_structure:
            new_files.extend([os.path.join(dir_path, file) for file in files])
        else:
            new_files.extend([os.path.join(dir_path, file) for file in files if file not in old_structure[dir_path]])

    for dir_path, files in old_structure.items():
        if dir_path not in new_structure:
            moved_files.extend([os.path.join(dir_path, file) for file in files])
        else:
            moved_files.extend([os.path.join(dir_path, file) for file in files if file not in new_structure[dir_path]])

    with open(log_file, 'a') as f:
        f.write(f"\nChanges detected on {datetime.now()}:\n")
        for file in new_files:
            f.write(f"NewFile: {file}\n")
        for file in moved_files:
            f.write(f"MovedFile: {file}\n")

def main():
    new_structure = get_file_structure(ROOT_DIR)
    save_file_structure(new_structure, OUTPUT_FILE)

    if os.path.exists(PREVIOUS_STRUCTURE_FILE):
        with open(PREVIOUS_STRUCTURE_FILE, 'r') as f:
            old_structure = json.load(f)
        log_changes(new_structure, old_structure, LOG_FILE)

    with open(PREVIOUS_STRUCTURE_FILE, 'w') as f:
        json.dump(new_structure, f)

if __name__ == "__main__":
    main() 